### ISSUE TITLE
GradCAM and nested models

### ISSUE THREAD

**Issue by palatos** [tensorflow/tensorflow#44462]
- TensorFlow version: 2.3.0
- Python version: 3.7.6

I'm trying to reproduce the Keras GradCAM example (https://keras.io/examples/vision/grad_cam/) using a custom model.
My custom model is relatively simple and inspired by the TF/Keras fine tuning tutorial (https://www.tensorflow.org/guide/keras/transfer_learning).

The problem is, adding a model such as vgg16 with the functional API, as a layer to my main model, seems to "break the graph" and not allow part of the GradCAM tutorial code to work properly. The model itself works fine for prediction and etc, but upon trying to create a 'last_conv_layer_model' I keep getting a "Graph disconnected" error.

I have tried flattening the model, so that vgg16 is not nested within the main model anymore, but this didn't change the error.

I've seen many people trying to figure out this problem and apparently the only solution is to never use the functional API in this way (allowing for nested models to happen), but that seems like a problem especially for already trained models.  I thought I'd create an issue here since maybe someone has a more elegant solution.

I can't upload my actual model but I made a notebook that reproduces the error and all of the described process with a similar architecture. 

https://github.com/palatos/mynotes/blob/main/gradcam-with-nested-models-error.ipynb

(The model is not fine tuned yet in the notebook but that doesn't matter. We could generate GradCAM heatmaps even for gibberish, non-fine tuned, predictions)

Thanks in advance for any clarification on this issue.

**Comment by Saduf2019** [comment]
I am able to replicate the issue, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/00393cd5c9f62227595cc0c6569c2ac1/untitled458.ipynb)

**Comment by nikitamaia** [comment]
Adding the `contributions welcome` label to this issue for further investigation by the community. If you are interested in working on this issue, please leave a comment and I will assign it to you. Thanks!

**Comment by bhack** [comment]
@palatos More then Gradcam specific, that is hosted on [another repository](https://github.com/keras-team/keras-io/blob/master/examples/vision/grad_cam.py), is in your specific case more related to  https://github.com/tensorflow/tensorflow/issues/34977 ?

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210530, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/1aa531794cc8e01a8c8763069e30881c/44462.ipynb). Thanks!

**Comment by IEPiXI** [comment]
Are there any new findings/solutions to this problem yet? I'm encountering the same issue and could need some help.

**Comment by palatos** [comment]
Just to give an update.
I wasn't able to successfully fix the problem with the get_output_at method in this case, but if someone manages to do it please share the solution.

I have a paliative solution for this issue but it's not very elegant and it requires a lot of manual changes for each specific nested model situation. I've uploaded the solution to this notebook:

https://github.com/palatos/mynotes/blob/main/gradcam_paliative.ipynb

Essentially what changes is that I manually separate the preprocessing, convolutional and classification blocks while generating the heatmaps, avoiding the whole ambiguity with inputs and outputs. That way I can preprocess the image before we begin, then pass it only through the convolutional layers, and then pass the result of that only through the classification layers. This eliminates the problem, but requires a lot of changes if you want to adapt it to your particular problem.

```

#Make GradCAM heatmap following the Keras tutorial.
last_conv_layer = model.layers[-4].layers[-1]
last_conv_layer_model = keras.Model(model.layers[-4].inputs, last_conv_layer.output)

# Second, we create a model that maps the activations of the last conv
# layer to the final class predictions
classifier_input = keras.Input(shape=last_conv_layer.output.shape[1:])
x = classifier_input
for layer in model.layers[-3:]:
    x = layer(x)
classifier_model = keras.Model(classifier_input, x)

#Preparing the image with the preprocessing layers
preprocess_layers = keras.Model(model.inputs, model.layers[-5].output)
img_array = preprocess_layers(prepared_image)

# Then, we compute the gradient of the top predicted class for our input image
# with respect to the activations of the last conv layer
with tf.GradientTape() as tape:
    # Compute activations of the last conv layer and make the tape watch it
    last_conv_layer_output = last_conv_layer_model(img_array)
    tape.watch(last_conv_layer_output)
    # Compute class predictions
    preds = classifier_model(last_conv_layer_output)
    top_pred_index = tf.argmax(preds[0])
    top_class_channel = preds[:, top_pred_index]

# This is the gradient of the top predicted class with regard to
# the output feature map of the last conv layer
grads = tape.gradient(top_class_channel, last_conv_layer_output)

```

From what I read I should be able to get the same result in a much more elegant way by using get_output_at(), but I wasn't able to.

**Comment by rchao** [comment]
@fchollet cc'ing you for your thoughts.

**Comment by EddyJens** [comment]
Same problem here!
using:
tensorflow 2.8
python 3.8

After reading many other people in google, I believe that the solution is to somehow turn the model flat;

the code above kinda does that, although the result of it (at least in my model) was super weird - despite having a good accuracy, the corresponding grad-cam heatmap was showing the lower left and right corner as more relevant pixels to the image classification (that is not good, because the object being classified is at the center of the image).

I tried also with a different model, but the result was in the same way suspect, because now the heatmap was  only 3 pixels disposed near the center of the image.
